File: dbos_server_service/web_ui/src/api/secret/roleAcls.ts
/**
 * Thin wrappers для `secret_service` `/secret/v1/credentials/{id}/acl`.
 *
 * Source of truth:
 *   secret_service/src/api/v1/endpoints/role_acls.py
 *   secret_service/src/schemas/role_acls.py
 */

import { apiDelete, apiGet, apiPost, apiPut } from "@/api/client";
import type {
  OkResponse,
  RoleACL,
  RoleACLCreateRequest,
  RoleACLList,
  RoleACLUpsertRequest,
  RoleACLUpsertResponse,
} from "@/api/secret/types";

const BASE = "/secret/v1/credentials";

/** `GET /credentials/{id}/acl` — список RoleACL кред. */
export function listRoleAcls(credId: string): Promise<RoleACLList> {
  return apiGet<RoleACLList>(`${BASE}/${credId}/acl`);
}

/**
 * `POST /credentials/{id}/acl` — выдать RoleACL.
 *
 * `dept_id` + `role_name` обязательны; cross_department recipient требует
 * заранее выданного DeptGrant, иначе backend отбивает.
 */
export function addRoleAcl(
  credId: string,
  body: RoleACLCreateRequest,
): Promise<RoleACL> {
  return apiPost<RoleACL>(`${BASE}/${credId}/acl`, body);
}

/**
 * `PUT /credentials/{id}/acl` — атомарный upsert RoleACL.
 *
 * Задаёт желаемую пару `(can_read, can_write)` для `(dept_id, role_name)`:
 * строки нет — создаёт, есть — переписывает флаги, оба флага false — снимает
 * строку. Идемпотентно, без 409 — заменяет связку revoke+re-add для тоггла
 * ячейки матрицы. `acl` в ответе = null, когда доступ снят.
 */
export function upsertRoleAcl(
  credId: string,
  body: RoleACLUpsertRequest,
): Promise<RoleACLUpsertResponse> {
  return apiPut<RoleACLUpsertResponse>(`${BASE}/${credId}/acl`, body);
}

/** `DELETE /credentials/{id}/acl/{aclId}` — снять RoleACL. */
export function revokeRoleAcl(
  credId: string,
  aclId: string,
): Promise<OkResponse> {
  return apiDelete<OkResponse>(`${BASE}/${credId}/acl/${aclId}`);
}

File: dbos_server_service/secret_service/src/api/v1/endpoints/role_acls.py
"""HTTP-эндпоинты /credentials/{id}/acl — RoleACL CRUD."""

from fastapi import APIRouter, Depends, Request, status
from sqlalchemy.ext.asyncio import AsyncSession

from src.core.config import get_settings
from src.core.limiter import limiter
from src.dependencies.auth import CurrentIdentity, require_user_context
from src.dependencies.db import get_db
from src.models import RoleACL
from src.schemas.common import OkResponse
from src.schemas.role_acls import (
    RoleACLCreate,
    RoleACLList,
    RoleACLRead,
    RoleACLUpsert,
    RoleACLUpsertResponse,
)
from src.services import role_acl_service

router = APIRouter(prefix="/credentials", tags=["role_acls"])


def _to_read(acl: RoleACL) -> RoleACLRead:
    return RoleACLRead(
        id=acl.id,
        cred_id=acl.cred_id,
        dept_id=acl.dept_id,
        role_name=acl.role_name,
        can_read=acl.can_read,
        can_write=acl.can_write,
        granted_by_user_id=acl.granted_by_user_id,
        granted_at=acl.granted_at,
    )


@router.post(
    "/{cred_id}/acl",
    response_model=RoleACLRead,
    status_code=status.HTTP_201_CREATED,
    summary="Выдать RoleACL (cross_department recipient требует DeptGrant)",
)
@limiter.limit(get_settings().rate_limit_acl)
async def add_acl(
    cred_id: str,
    request: Request,
    payload: RoleACLCreate,
    identity: CurrentIdentity,
    db: AsyncSession = Depends(get_db),
) -> RoleACLRead:
    require_user_context(identity)
    acl = await role_acl_service.add(db, identity, cred_id, payload)
    return _to_read(acl)


@router.put(
    "/{cred_id}/acl",
    response_model=RoleACLUpsertResponse,
    summary="Upsert RoleACL: задать пару (can_read, can_write) для (dept, role)",
)
@limiter.limit(get_settings().rate_limit_acl)
async def upsert_acl(
    cred_id: str,
    request: Request,
    payload: RoleACLUpsert,
    identity: CurrentIdentity,
    db: AsyncSession = Depends(get_db),
) -> RoleACLUpsertResponse:
    require_user_context(identity)
    acl = await role_acl_service.upsert(db, identity, cred_id, payload)
    return RoleACLUpsertResponse(ok=True, acl=_to_read(acl) if acl else None)


@router.get(
    "/{cred_id}/acl",
    response_model=RoleACLList,
    summary="Список RoleACL для credential",
)
async def list_acls(
    cred_id: str,
    identity: CurrentIdentity,
    db: AsyncSession = Depends(get_db),
) -> RoleACLList:
    require_user_context(identity)
    acls = await role_acl_service.list_for(db, identity, cred_id)
    return RoleACLList(items=[_to_read(a) for a in acls])


@router.delete(
    "/{cred_id}/acl/{acl_id}",
    response_model=OkResponse,
    summary="Снять RoleACL",
)
async def revoke_acl(
    cred_id: str,
    acl_id: str,
    identity: CurrentIdentity,
    db: AsyncSession = Depends(get_db),
) -> OkResponse:
    require_user_context(identity)
    await role_acl_service.revoke(db, identity, cred_id, acl_id)
    return OkResponse(ok=True)
